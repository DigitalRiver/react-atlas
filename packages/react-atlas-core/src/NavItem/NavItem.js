import React from "react";
import PropTypes from "prop-types";
import cx from "classnames";
import { ButtonCore } from "./../Button";

class NavItem extends React.PureComponent {
  _handleClick = event => {
    if (this.props.onClick) {
      this.props.onClick(this.props.navKey, event);
    }
  };

  render() {
    const {
      className,
      style,
      active,
      disabled,
      href,
      parent,
      subNav,
      collapsed,
      children
    } = this.props;
    if (subNav && collapsed) {
      return "";
    }

    return (
      <li
        role="presentation"
        styleName={cx("navItem", { active, disabled, collapsed })}
        style={style}
        className={className}
      >
        <ButtonCore
          disabled={disabled}
          styleName={cx("link", { disabled, subNav })}
          href={href}
          onClick={this._handleClick}
        >
          {children}
          <i styleName={cx({ "caret": parent })} />
        </ButtonCore>
      </li>
    );
  }
}

NavItem.propTypes = {
  /**
   * Child elements, plain text or HTML/React element.
   */
  "children": PropTypes.node.isRequired,

  /** An object, array, or string of CSS classes to apply to NavItem.*/
  "className": PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.object,
    PropTypes.array
  ]),

  /**
   * When true, NavItem will be disabled.
   */
  "disabled": PropTypes.bool,

  /**
   * The URL that the NavItem will link to.
   */
  "href": PropTypes.string,

  /**
   * Specify NavItem unique key
   */
  "navKey": PropTypes.number.isRequired,
  /**
   * Pass inline styling here.
   */
  "style": PropTypes.object,
  /**
   * When true, NavItem will be set as active.
   *
   * @ignore
   */
  "active": PropTypes.bool,
  /**
   * Only first NavItem in sub-Nav container will be true.
   *
   * @ignore
   */
  "parent": PropTypes.bool,
  /**
   * Each NavItem in sub-Nav container will be true except first one.
   *
   * @ignore
   */
  "subNav": PropTypes.bool,
  /**
   * Only use in sub Nav container, defines parent NavItem caret and subNav show or not.
   *
   * @ignore
   */
  "collapsed": PropTypes.bool,
  /**
   * callback for Nav Component
   *
   * @ignore
   */
  "onClick": PropTypes.func
};

NavItem.defaultProps = {
  "className": ""
};

export default NavItem;
