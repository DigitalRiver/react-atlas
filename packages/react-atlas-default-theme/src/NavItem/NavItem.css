@import '../Button/Button.css';

.navItem button {
    width: 100%;
    border-color: #eee;
    border-left: solid 5px transparent;
    color: #555;
    position: relative;
    display: block;
    padding: 10px 15px;
    background-color: #fff;
    text-align: left;
}

.navItem button:hover {
    background-color: #eee;
}

.active > button {
    border-left-color: #003058;
    background-color: #eee;
    color: #003058;
}

.base {
    border-style: none;
    border-top: 1px solid #ccc;
}

.navItem .caret {
    position: relative;
    display: inline-block;
    width: 18px;
    height: 10px;
    vertical-align: middle;
    float: right;
    clear: right;
    border-color: #555;
    transform: none;
    margin: 5px 0 0;
    transition: transform 300ms;
}

.collapsed .caret {
    transform: rotate(90deg);
    transition: transform 300ms;
}

.navItem .caret::before {
    content: "";
    border: solid #003058;
    border-width: 0 2px 2px 0;
    display: inline-block;
    padding: 5px;
    transform: rotate(45deg);
    position: absolute;
    top: -5px;
    left: 5px;
}

.navItem .subNav {
    padding-left: 30px;
}