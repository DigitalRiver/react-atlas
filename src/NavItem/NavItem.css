@import '../Button/Button.css';
@import '../styles/variables.css';

.base {
    border-style: none;
    /*border-top: 1px solid #eee !important; !* Important designation needed to overwrite Button styling in prod build *!*/
}

.subNav {
    height: 40px;
    transition: height .3s ease-in-out;
}

.subNav.collapsed.initial {
    height: auto;
}

.subNav.collapsed.initial button {
    display: list-item;
    visibility: hidden;
}

.subNav.collapsed.normal {
    height: 0;
}

.subNav.collapsed.normal button {
    display: none;
    visibility: visible;
    transition: height .3s ease-in-out;
}

.navItem button {
    width: 100%;
    border-color: var(--silver);
    border-left: solid 5px transparent;
    color: var(--dark-gray);
    position: relative;
    display: block;
    padding: 10px 15px;
    background-color: var(--white);
    text-align: left;
}

.horizontal button {
    border-top: solid 5px transparent;
    border-left: none;
}

.horizontal ~ li > button {
    padding-left: 10px !important;
}

.subNav button {
    padding-left: 38px;
}

.navItem button:hover {
    background-color: var(--silver);
    color: var(--dark-gray);
}

.navItem button:focus {
    outline: none;
}

.active > button, .active > a > button {
    border-left-color: var(--blue);
    background-color: var(--silver);
    color: var(--blue);
}

.horizontal.active > button {
    border-top: solid 5px var(--blue) !important;
}

.caret::before {
    content: "";
    border: solid var(--blue);
    border-width: 0 2px 2px 0;
    display: inline-block;
    padding: 5px;
    position: absolute;
    top: -5px;
    left: 5px;
}

.caretHorizontal::before {
    transform: rotate(45deg);
}

.caretVertical::before {
    transform: rotate(45deg);
}

.caret {
    position: relative;
    width: 18px;
    height: 10px;
    vertical-align: middle;
    float: right;
    clear: right;
    border-color: var(--dark-gray);
    transform: none;
    margin: 5px 0 0;
    transition: transform .3s;
    top: -2px;
}


.collapsed .caretVertical {
    transform: rotate(90deg);
}