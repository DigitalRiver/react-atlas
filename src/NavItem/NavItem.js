import React from "react";
import PropTypes from "prop-types";
import cx from "classnames";
import { Button } from "../Button";
import CSSModules from "react-css-modules";
import styles from "./NavItem.css";

export class NavItem extends React.Component {
  constructor(props) {
    super(props);
    this.liRef = React.createRef();
    this.state = {
      "initial": true
    };
  }

  componentDidMount() {
    if (this.liRef.current && this.props.horizontal && this.props.subNav) {
      const div = window.getComputedStyle(this.liRef.current);
      const navItemHeight = parseInt(
        div
          .getPropertyValue("height")
          .substring(0, div.getPropertyValue("height").length - 2)
      );
      const navItemWidth = parseInt(
        div
          .getPropertyValue("width")
          .substring(0, div.getPropertyValue("width").length - 2)
      );
      this.props.dimension(navItemHeight, navItemWidth);
    }
    this.setState({
      "initial": false
    });
  }

  _handleClick = event => {
    if (this.props.onClick) {
      this.props.onClick(this.props.navKey, event);
    }
  };

  render() {
    const {
      as,
      className,
      style,
      active,
      disabled,
      href,
      parent,
      subNav,
      collapsed,
      children,
      horizontal,
      ...others
    } = this.props;

    const styleNameHelper = (horizontalCond, subNavCond, initialCond) => {
      if (!horizontalCond || horizontalCond && subNavCond) {
        if (initialCond) {
          return cx("navItem", "initial", {
            active,
            disabled,
            collapsed,
            subNav
          });
        } else {
          return cx("navItem", "normal", {
            active,
            disabled,
            collapsed,
            subNav
          });
        }
      } else {
        if (initialCond) {
          return cx("navItem", "initial", "horizontal", {
            active,
            disabled,
            collapsed,
            subNav
          });
        } else {
          return cx("navItem", "normal", "horizontal", {
            active,
            disabled,
            collapsed,
            subNav
          });
        }
      }
    };

    const button = 
      <Button
        disabled={disabled}
        styleName={
          !horizontal || horizontal && subNav
            ? cx("link", disabled, subNav)
            : cx("link", "horizontal", disabled, subNav)
        }
        link
        href={typeof to === "undefined" ? href : null}
        onClick={typeof to === "undefined" ? this._handleClick : null}
      >
        {children}
        <i
          styleName={cx({
            "caret": parent,
            "caretVertical": !horizontal,
            "caretHorizontal": horizontal
          })}
        />
      </Button>
    ;

    const LinkElement = this.props.as;

    if (typeof as !== "undefined" && !disabled) {
      return (
        <li
          role="presentation"
          // keep the comment in case need to change it back

          // styleName={
          //   !horizontal || horizontal && subNav
          //     ? cx("navItem", { active, disabled, collapsed, subNav })
          //     : cx("navItem", "horizontal", {
          //         active,
          //         disabled,
          //         collapsed,
          //         subNav
          //       })
          // }
          styleName={styleNameHelper(horizontal, subNav, this.state.initial)}
          style={style}
          className={className}
          ref={parent && horizontal ? this.liRef : null}
        >
          <LinkElement {...others} onClick={this._handleClick}>
            {button}
          </LinkElement>
        </li>
      );
    } else {
      return (
        <li
          role="presentation"
          // keep the comment in case need to change it back

          // styleName={
          //   !horizontal || horizontal && subNav
          //     ? cx("navItem", { active, disabled, collapsed, subNav })
          //     : cx("navItem", "horizontal", {
          //         active,
          //         disabled,
          //         collapsed,
          //         subNav
          //       })
          // }
          styleName={styleNameHelper(horizontal, subNav, this.state.initial)}
          style={style}
          className={className}
          ref={!parent && horizontal ? this.liRef : null}
        >
          {button}
        </li>
      );
    }
  }
}

NavItem.propTypes = {
  /** Child elements, plain text or HTML/React element. */
  "as": PropTypes.func,

  /** Child elements, plain text or HTML/React element. */
  "children": PropTypes.node.isRequired,

  /** An object, array, or string of CSS classes to apply to NavItem.*/
  "className": PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.object,
    PropTypes.array
  ]),

  /** When true, NavItem will be disabled. */
  "disabled": PropTypes.bool,

  /** The URL that the NavItem will link to. */
  "href": PropTypes.string,

  /** Specify NavItem unique key */
  "navKey": PropTypes.number.isRequired,
  /** Pass inline styling here. */
  "style": PropTypes.object,
  /** When true, NavItem will be set as active.
   * @ignore
   */
  "active": PropTypes.bool,
  /** Only first NavItem in sub-Nav container will be true.
   * @ignore
   */
  "parent": PropTypes.bool,
  /** Each NavItem in sub-Nav container will be true except first one.
   * @ignore
   */
  "subNav": PropTypes.bool,
  /** Only use in sub Nav container, defines parent NavItem caret and subNav show or not.
   * @ignore
   */
  "collapsed": PropTypes.bool,
  /** callback for Nav Component
   * @ignore
   */
  "onClick": PropTypes.func,
  /** Define whether the NavItem is vertical or horizontal, NavItem is vertical by default */
  "horizontal": PropTypes.bool,
  /** callback for Nav Component for the height and width of the NavItem */
  "dimension": PropTypes.func
};

NavItem.defaultProps = {
  "className": "",
  "horizontal": false
};

export default CSSModules(NavItem, styles, { "allowMultiple": true });
