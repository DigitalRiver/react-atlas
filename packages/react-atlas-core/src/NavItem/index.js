import { default as NavItemCore } from "./NavItem";
export { NavItemCore };
