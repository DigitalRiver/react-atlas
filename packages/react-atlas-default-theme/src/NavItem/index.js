import * as NavItemStyle from "./NavItem.css";
export { NavItemStyle };
