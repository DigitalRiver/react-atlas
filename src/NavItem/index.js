import NavItem from "./NavItem.js";
export { NavItem };
