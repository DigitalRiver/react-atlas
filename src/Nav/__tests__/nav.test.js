import React from "react";
import { mount } from "enzyme";
import { Nav } from "../index";
import { NavItem } from "../../NavItem";
import renderer from "react-test-renderer";
import data from "./data";

describe("Test Nav component render", () => {
  it("Render correctly", () => {
    const tree = renderer
      .create(
        <Nav>
          <NavItem navKey={0}>HOME</NavItem>
          <NavItem navKey={1}>Catalog</NavItem>
          <NavItem navKey={2}>Customer Service</NavItem>
          <NavItem navKey={3}>Administration</NavItem>
          <NavItem navKey={4}>Support</NavItem>
        </Nav>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});

const _findItem = (node, navKey) =>
  node.name() === "NavItem" && node.props().navKey === navKey;

const _findActiveItem = node =>
  node.name() === "NavItem" && node.props().active;

describe("Test Nav component - Basic tests", () => {
  it("Nav - Simple default select NavItem", function() {
    const result = mount(
      <Nav activeIndex={1}>
        <NavItem navKey={0}>HOME</NavItem>
        <NavItem navKey={1}>Catalog</NavItem>
        <NavItem navKey={2}>Customer Service</NavItem>
        <NavItem navKey={3}>Administration</NavItem>
        <NavItem navKey={4}>Support</NavItem>
      </Nav>
    );
    expect(result.findWhere(n => _findActiveItem(n)).props().navKey).toEqual(1);
  });
  it("Nav - Default select inner NavItem", function() {
    const result = mount(
      <Nav activeIndex={3.1}>
        <NavItem navKey={0}>HOME</NavItem>
        <NavItem navKey={1}>Catalog</NavItem>
        <Nav>
          <NavItem navKey={3}>Reports</NavItem>
          <NavItem navKey={3.1}>Create a Report</NavItem>
          <NavItem navKey={3.2}>Edit a Report</NavItem>
        </Nav>
        <NavItem navKey={4}>Administration</NavItem>
        <NavItem navKey={5}>Support</NavItem>
      </Nav>
    );
    expect(result.findWhere(n => _findActiveItem(n)).props().navKey).toEqual(
      3.1
    );
  });

  it("Nav - Click to changing active NavItem", function() {
    const result = mount(
      <Nav activeIndex={1}>
        <NavItem navKey={0}>HOME</NavItem>
        <NavItem navKey={1}>Catalog</NavItem>
        <NavItem navKey={2}>Customer Service</NavItem>
        <Nav>
          <NavItem navKey={3}>Reports</NavItem>
          <NavItem navKey={3.1}>Create a Report</NavItem>
          <NavItem navKey={3.2}>Edit a Report</NavItem>
        </Nav>
        <NavItem navKey={4}>Support</NavItem>
      </Nav>
    );
    result
      .findWhere(n => _findItem(n, 2))
      .find("button")
      .simulate("click");
    expect(result.findWhere(n => _findActiveItem(n)).props().navKey).toEqual(2);

    result
      .findWhere(n => _findItem(n, 3.1))
      .find("button")
      .simulate("click");
    expect(result.findWhere(n => _findActiveItem(n)).props().navKey).toEqual(
      3.1
    );
  });

  it("Nav - handle collapse", function() {
    const nav = mount(
      <Nav activeIndex={1}>
        <NavItem navKey={0}>HOME</NavItem>
        <NavItem navKey={1}>Catalog</NavItem>
        <NavItem navKey={2}>Customer Service</NavItem>
        <Nav className="myNav">
          <NavItem navKey={3}>Reports</NavItem>
          <NavItem navKey={3.1}>Create a Report</NavItem>
          <NavItem navKey={3.2}>Edit a Report</NavItem>
        </Nav>
        <NavItem navKey={4}>Support</NavItem>
      </Nav>
    );
    const myNav = nav.find(".myNav").first();
    expect(myNav.instance().state.collapsed).toBe(false);
    nav
      .find(".myNav")
      .first()
      .find("button")
      .first()
      .simulate("click");
    expect(myNav.instance().state.collapsed).toBe(true);
  });
});

describe("Test JSON data to rendering Nav component", () => {
  it("Simple default select NavItem", () => {
    const result = mount(<Nav activeIndex={2} data={data} />);
    expect(result.findWhere(n => _findActiveItem(n)).props().navKey).toEqual(2);
  });
});
