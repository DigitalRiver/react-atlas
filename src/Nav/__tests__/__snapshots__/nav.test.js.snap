// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Nav component render Render correctly 1`] = `
<ul
  className="nav"
  style={undefined}
>
  <li
    className="navItem"
    role="presentation"
    style={undefined}
  >
    <button
      className="link button link link_hover"
      href={undefined}
      id={undefined}
      name={undefined}
      onClick={[Function]}
      style={undefined}
      tabIndex={0}
      type="button"
    >
      HOME
      <i />
    </button>
  </li>
  <li
    className="navItem"
    role="presentation"
    style={undefined}
  >
    <button
      className="link button link link_hover"
      href={undefined}
      id={undefined}
      name={undefined}
      onClick={[Function]}
      style={undefined}
      tabIndex={0}
      type="button"
    >
      Catalog
      <i />
    </button>
  </li>
  <li
    className="navItem"
    role="presentation"
    style={undefined}
  >
    <button
      className="link button link link_hover"
      href={undefined}
      id={undefined}
      name={undefined}
      onClick={[Function]}
      style={undefined}
      tabIndex={0}
      type="button"
    >
      Customer Service
      <i />
    </button>
  </li>
  <li
    className="navItem"
    role="presentation"
    style={undefined}
  >
    <button
      className="link button link link_hover"
      href={undefined}
      id={undefined}
      name={undefined}
      onClick={[Function]}
      style={undefined}
      tabIndex={0}
      type="button"
    >
      Administration
      <i />
    </button>
  </li>
  <li
    className="navItem"
    role="presentation"
    style={undefined}
  >
    <button
      className="link button link link_hover"
      href={undefined}
      id={undefined}
      name={undefined}
      onClick={[Function]}
      style={undefined}
      tabIndex={0}
      type="button"
    >
      Support
      <i />
    </button>
  </li>
</ul>
`;
